src/hooks/useConversation.ts

```diff
@@ -145,14 +147,29 @@ export function useConversation(conversation: ConversationItem): UseConversation
         onMessageAdded: (message) => {
           console.log('[useConversation] Received message:', message);
 
-          // if from assistant, don't output since we already got it
-          if (message.role === 'assistant') return;
-
           // Update the conversation with the message
           queryClient.setQueryData<ConversationResponse>(
             ['conversation', conversation.name, conversation.readonly],
             (old) => {
               if (!old) return undefined;
+              // Check last 2 messages for duplicates
+              // Prevents duplicate messages from being added to the log, as in the cases of:
+              //  - onMessageAdded running after onComplete (effectively a no-op)
+              //  - message_added events for messages we sent outselves (already added to conversation)
+              const recentMessages = old.log.slice(-2);
+              const isDuplicate = recentMessages.some(
+                (msg) => msg.role === message.role && msg.content === message.content
+                // && msg.timestamp === message.timestamp
+              );
```

Comment:
This is not so robust as the user can send two separate messages with the same content. Better to add back message.id